be/src/exec/pipeline/exchange/mem_limited_chunk_queue.cpp

```diff
@@ -0,0 +1,528 @@
+            "ExchangerPeakMemoryUsage", TUnit::BYTES,
+            RuntimeProfile::Counter::create_strategy(TUnit::BYTES, TCounterMergeType::SKIP_FIRST_MERGE));
+    _peak_memory_rows_counter = parent->AddHighWaterMarkCounter(
+            "ExchangerPeakBufferRowSize", TUnit::UNIT,
+            RuntimeProfile::Counter::create_strategy(TUnit::UNIT, TCounterMergeType::SKIP_FIRST_MERGE));
+
+    _flush_io_timer = ADD_TIMER(parent, "FlushIOTime");
+    _flush_io_count = ADD_COUNTER(parent, "FlushIOCount", TUnit::UNIT);
+    _flush_io_bytes = ADD_COUNTER(parent, "FlushIOBytes", TUnit::BYTES);
+    _read_io_timer = ADD_TIMER(parent, "ReadIOTime");
+    _read_io_count = ADD_COUNTER(parent, "ReadIOCount", TUnit::UNIT);
+    _read_io_bytes = ADD_COUNTER(parent, "ReadIOBytes", TUnit::BYTES);
+    return Status::OK();
+}
+
+Status MemLimitedChunkQueue::push(const ChunkPtr& chunk) {
+    RETURN_IF_ERROR(_get_io_task_status());
+    std::unique_lock l(_mutex);
+    if (UNLIKELY(_chunk_builder == nullptr)) {
+        _chunk_builder = chunk->clone_empty();
+    }
+    int32_t closed_source_number = _consumer_number - _opened_source_number;
+
+    DCHECK(_tail != nullptr) << "tail won't be null";
+    DCHECK(_tail->next == nullptr) << "tail should be the last block";
+
+    // create a new block
+    if (_tail->memory_usage >= _opts.block_size) {
+        Block* block = new Block();
+        block->next = nullptr;
+        _tail->next = block;
+        _tail = block;
+    }
+
+    _total_accumulated_rows += chunk->num_rows();
+    _total_accumulated_bytes += chunk->memory_usage();
+
+    Cell cell;
+    cell.chunk = chunk;
+    cell.used_count = closed_source_number;
+    cell.accumulated_rows = _total_accumulated_rows;
+    cell.accumulated_bytes = _total_accumulated_bytes;
+    _tail->cells.emplace_back(cell);
+    _tail->memory_usage += chunk->memory_usage();
+
+    size_t in_memory_rows = _total_accumulated_rows - _flushed_accumulated_rows + _current_load_rows;
+    size_t in_memory_bytes = _total_accumulated_bytes - _flushed_accumulated_bytes + _current_load_bytes;
+
+#ifndef BE_TEST
+    _peak_memory_rows_counter->set(in_memory_rows);
+    _peak_memory_bytes_counter->set(in_memory_bytes);
+#endif
+    return Status::OK();
+}
+
+bool MemLimitedChunkQueue::can_push() {
+    std::shared_lock l(_mutex);
+    size_t unconsumed_bytes = _total_accumulated_bytes - _fastest_accumulated_bytes;
+    // if the fastest consumer still has a lot of data to consume, it will no longer accept new input.
+    if (unconsumed_bytes >= _opts.max_unconsumed_bytes) {
+        return false;
+    }
+
+    size_t in_memory_bytes = _total_accumulated_bytes - _flushed_accumulated_bytes;
+    if (in_memory_bytes >= _opts.memory_limit && _next_flush_block->next != nullptr) {
+        if (bool expected = false; _has_flush_io_task.compare_exchange_strong(expected, true)) {
+            TEST_SYNC_POINT("MemLimitedChunkQueue::can_push::before_submit_flush_task");
```

Comment:
Only one pending io task is allowed? how about launch multiple io tasks concurrently?

Reply:
> Only one pending io task is allowed? how about launch multiple io tasks concurrently?

I'm not sure whether there will always be benefits to doing so, as we cannot predict the reading of each consumer. If too many blocks are flushed at once, more data may need to be loaded from disk during reading.
I think we can consider doing more optimizations in the future, but now may not be a very critical issue